apps/dashboard/pkg/migration/schemaversion/v30.go

```diff
@@ -0,0 +1,393 @@
+// - Maintains range mappings as separate RangeToText mappings
+//
+// Tooltip Options Migration:
+// - Renames panel.options.tooltipOptions to panel.options.tooltip
+// - Only applies to timeseries, xychart, and xychart2 panel types
+// - Preserves all tooltip configuration options
+//
+// Example value mappings transformation:
+//
+// Before migration:
+//
+//	panel: {
+//	  "fieldConfig": {
+//	    "defaults": {
+//	      "mappings": [
+//	        { "id": 0, "text": "Up", "type": 1, "value": "1" },
+//	        { "id": 1, "text": "Down", "type": 1, "value": "0" },
+//	        { "from": "10", "to": "20", "text": "Medium", "type": 2 }
+//	      ]
+//	    }
+//	  }
+//	}
+//
+// After migration:
+//
+//	panel: {
+//	  "fieldConfig": {
+//	    "defaults": {
+//	      "mappings": [
+//	        {
+//	          "type": "value",
+//	          "options": {
+//	            "1": { "text": "Up" },
+//	            "0": { "text": "Down" }
+//	          }
+//	        },
+//	        {
+//	          "type": "range",
+//	          "options": {
+//	            "from": 10,
+//	            "to": 20,
+//	            "result": { "text": "Medium" }
+//	          }
+//	        }
+//	      ]
+//	    }
+//	  }
+//	}
+//
+// Example tooltip options transformation:
+//
+// Before migration:
+//
+//	panel: {
+//	  "type": "timeseries",
+//	  "options": {
+//	    "tooltipOptions": { "mode": "multi" }
+//	  }
+//	}
+//
+// After migration:
+//
+//	panel: {
+//	  "type": "timeseries",
+//	  "options": {
+//	    "tooltip": { "mode": "multi" }
+//	  }
+//	}
+func V30(dashboard map[string]interface{}) error {
+	dashboard["schemaVersion"] = 30
+
+	panels, ok := dashboard["panels"].([]interface{})
+	if !ok {
+		return nil
+	}
+
+	for _, panel := range panels {
+		panelMap, ok := panel.(map[string]interface{})
+		if !ok {
+			continue
+		}
+
+		upgradeValueMappingsForPanel(panelMap)
+		migrateTooltipOptions(panelMap)
+
+		// Handle nested panels in collapsed rows
+		nestedPanels, hasNested := panelMap["panels"].([]interface{})
+		if !hasNested {
+			continue
+		}
+
+		for _, nestedPanel := range nestedPanels {
+			nestedPanelMap, ok := nestedPanel.(map[string]interface{})
+			if !ok {
+				continue
+			}
+			upgradeValueMappingsForPanel(nestedPanelMap)
+			migrateTooltipOptions(nestedPanelMap)
+		}
+	}
+
+	return nil
+}
+
+// upgradeValueMappingsForPanel migrates value mappings from old format to new format
+func upgradeValueMappingsForPanel(panel map[string]interface{}) {
+	fieldConfig, ok := panel["fieldConfig"].(map[string]interface{})
+	if !ok {
+		return
+	}
+
+	// Upgrade defaults mappings
+	if defaults, ok := fieldConfig["defaults"].(map[string]interface{}); ok {
+		if mappings, ok := defaults["mappings"].([]interface{}); ok {
+			var thresholds map[string]interface{}
+			if t, ok := defaults["thresholds"].(map[string]interface{}); ok {
+				thresholds = t
+			}
+			defaults["mappings"] = upgradeValueMappings(mappings, thresholds)
```

Comment:
do we want to add an empty one if the conversion fails?

Reply:
Yes Stephanie! This is what frontend does :) I added one snapshot for this case, so we ensure backend and frontend are consistent 👍